crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -655,4 +927,574 @@ mod tests {
+        let receipt_event_id = event_id!("$1");
+
+        let mut previous_events = Vector::new();
+
+        let ev1 = sync_timeline_message(other_user_id, receipt_event_id, "A");
+        let ev2 = sync_timeline_message(other_user_id, "$2", "A");
+
+        let receipt_event = EventBuilder::new().make_receipt_event_content([(
+            receipt_event_id.to_owned(),
+            ReceiptType::Read,
+            user_id.to_owned(),
+            ReceiptThread::Unthreaded,
+        )]);
+
+        let mut read_receipts = Default::default();
+        compute_notifications(
+            user_id,
+            room_id,
+            Some(&receipt_event),
+            previous_events.clone(),
+            &[ev1.clone(), ev2.clone()],
+            &mut read_receipts,
+        )
+        .unwrap();
+
+        // It did find the receipt event (ev1).
+        assert_eq!(read_receipts.num_unread, 1);
+
+        // Receive the same receipt event, with a new sync event.
+        previous_events.push_back(ev1);
+        previous_events.push_back(ev2);
+
+        let new_event = sync_timeline_message(other_user_id, "$3", "A");
+        compute_notifications(
+            user_id,
+            room_id,
+            Some(&receipt_event),
+            previous_events,
+            &[new_event],
+            &mut read_receipts,
+        )
+        .unwrap();
+
+        // Only the new event should be added.
+        assert_eq!(read_receipts.num_unread, 2);
+    }
+
+    fn make_test_events(user_id: &UserId) -> Vector<SyncTimelineEvent> {
+        let ev1 = sync_timeline_message(user_id, "$1", "With the lights out, it's less dangerous");
+        let ev2 = sync_timeline_message(user_id, "$2", "Here we are now, entertain us");
+        let ev3 = sync_timeline_message(user_id, "$3", "I feel stupid and contagious");
+        let ev4 = sync_timeline_message(user_id, "$4", "Here we are now, entertain us");
+        let ev5 = sync_timeline_message(user_id, "$5", "Hello, hello, hello, how low?");
+        vec![ev1, ev2, ev3, ev4, ev5].into()
+    }
+
+    /// Test that when multiple receipts come in a single event, we can still
+    /// find the latest one according to the sync order.
+    #[test]
+    fn test_compute_notifications_multiple_receipts_in_one_event() {
+        let user_id = user_id!("@alice:example.org");
+        let room_id = room_id!("!room:example.org");
+
+        let all_events = make_test_events(user_id!("@bob:example.org"));
+        let head_events: Vector<_> = all_events.iter().take(2).cloned().collect();
+        let tail_events: Vec<_> = all_events.iter().skip(2).cloned().collect();
+
+        for receipt_type_1 in &[ReceiptType::Read, ReceiptType::ReadPrivate] {
+            for receipt_thread_1 in &[ReceiptThread::Unthreaded, ReceiptThread::Main] {
+                for receipt_type_2 in &[ReceiptType::Read, ReceiptType::ReadPrivate] {
+                    for receipt_thread_2 in &[ReceiptThread::Unthreaded, ReceiptThread::Main] {
```

Comment:
Sometimes I find Given...When...Then comments help me understand my tests. Here it might be

```
// Given a receipt event marking events 1-3 as read using a combination of different thread and privacy types
```